File: src/dotnet/Harness.Engine/PlanObservationStore.cs
using System.Text.Json;

namespace Harness.Engine;

/// <summary>Append-only evidence that may justify a global plan revision.</summary>
public static class PlanObservationStore
{
    private const string Path = ".harness/plan_observations.jsonl";

    public static PlanObservation Append(string kind, int? featureId, string summary, params string[] evidence)
    {
        var observation = new PlanObservation(
            $"OBS-{Load().Count + 1:D3}", kind, featureId, summary,
            evidence.Where(e => !string.IsNullOrWhiteSpace(e)).ToArray(), DateTimeOffset.UtcNow);
        Directory.CreateDirectory(".harness");
        var json = JsonSerializer.Serialize(observation, HarnessJsonContext.Default.PlanObservation);
        File.AppendAllText(Path, json + Environment.NewLine);
        return observation;
    }

    public static IReadOnlyList<PlanObservation> Load()
    {
        if (!File.Exists(Path)) return [];
        var result = new List<PlanObservation>();
        foreach (var line in File.ReadLines(Path).Where(line => !string.IsNullOrWhiteSpace(line)))
        {
            try
            {
                if (JsonSerializer.Deserialize(line, HarnessJsonContext.Default.PlanObservation) is { } item)
                    result.Add(item);
            }
            catch { /* tolerate a partial final line after interruption */ }
        }
        return result;
    }

    public static void Reset()
    {
        try { if (File.Exists(Path)) File.Delete(Path); }
        catch (Exception ex) { HarnessLog.Error($"[PlanObservationStore] failed to reset: {ex.Message}"); }
    }
}

public sealed record PlanObservation(
    string Id,
    string Kind,
    int? FeatureId,
    string Summary,
    string[] Evidence,
    DateTimeOffset ObservedAt);

File: src/go/harnessengine/plan_observation_store.go
package harnessengine

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"
)

// Append-only evidence that may justify a global plan revision — persisted to
// .harness/plan_observations.jsonl (JSON Lines: one observation per line, never rewritten).
const planObservationsFilePath = ".harness/plan_observations.jsonl"

// PlanObservation is one piece of evidence (e.g. a feature failing deterministic
// verification repeatedly) recorded for later citation by a PlanRevision.
type PlanObservation struct {
	Id         string    `json:"id"`
	Kind       string    `json:"kind"`
	FeatureId  *int      `json:"featureId"`
	Summary    string    `json:"summary"`
	Evidence   []string  `json:"evidence"`
	ObservedAt time.Time `json:"observedAt"`
}

// AppendPlanObservation records a new observation and returns it. Id is
// "OBS-{running count + 1, zero-padded to 3 digits}" (e.g. "OBS-001").
func AppendPlanObservation(kind string, featureId *int, summary string, evidence ...string) PlanObservation {
	filtered := make([]string, 0, len(evidence))
	for _, e := range evidence {
		if strings.TrimSpace(e) != "" {
			filtered = append(filtered, e)
		}
	}

	observation := PlanObservation{
		Id:         fmt.Sprintf("OBS-%03d", len(LoadPlanObservations())+1),
		Kind:       kind,
		FeatureId:  featureId,
		Summary:    summary,
		Evidence:   filtered,
		ObservedAt: time.Now().UTC(),
	}

	if err := ensureDir(stateDir); err != nil {
		LogError(fmt.Sprintf("[PlanObservationStore] failed to append: %s", err))
		return observation
	}
	data, err := json.Marshal(observation)
	if err != nil {
		LogError(fmt.Sprintf("[PlanObservationStore] failed to append: %s", err))
		return observation
	}
	f, err := os.OpenFile(planObservationsFilePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		LogError(fmt.Sprintf("[PlanObservationStore] failed to append: %s", err))
		return observation
	}
	defer f.Close()
	if _, err := f.WriteString(string(data) + "\n"); err != nil {
		LogError(fmt.Sprintf("[PlanObservationStore] failed to append: %s", err))
	}
	return observation
}

// LoadPlanObservations reads every persisted observation, in append order. Tolerates a
// partial final line (e.g. after an interruption mid-write) by skipping it.
func LoadPlanObservations() []PlanObservation {
	if !fileExists(planObservationsFilePath) {
		return []PlanObservation{}
	}
	data, err := os.ReadFile(planObservationsFilePath)
	if err != nil {
		LogError(fmt.Sprintf("[PlanObservationStore] failed to load: %s", err))
		return []PlanObservation{}
	}

	result := []PlanObservation{}
	for _, line := range strings.Split(string(data), "\n") {
		if strings.TrimSpace(line) == "" {
			continue
		}
		var observation PlanObservation
		if err := json.Unmarshal([]byte(line), &observation); err == nil {
			result = append(result, observation)
		}
	}
	return result
}

// ResetPlanObservations clears the evidence log — paired with ResetFeatures/ResetPlanRevisions.
func ResetPlanObservations() {
	if !fileExists(planObservationsFilePath) {
		return
	}
	if err := os.Remove(planObservationsFilePath); err != nil {
		LogError(fmt.Sprintf("[PlanObservationStore] failed to reset: %s", err))
	}
}
